### PR TITLE
[0.60] Ensure WinRTWebSocketResource life time

### PR DESCRIPTION
**Pull request by JunielKatarn**
See #4330 .

Ensures solid references are kept when asynchronous methods that require the WebSocket resource are executed.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4465)

### CODE REVIEW

File: vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp [hunk 1]
```diff
@@ -45,34 +48,58 @@ return -1;
...
+  // TODO: Re-enable.
+  // server->Start();
 
   // WebSocket resources scope.
   {
     vector<shared_ptr<IWebSocketResource>> resources;
     for (int i = 0; i < resourceTotal; i++) {
-      auto ws = IWebSocketResource::Make("ws://localhost:5556/");
-      ws->SetOnMessage([this, &threadCount](size_t size, const string &message) {
+      auto ws = IWebSocketResource::Make("ws://localhost:5555/"); // TODO: Switch to port 5556
```
Comment (NickGerleman) [inline vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp:71]: Why the change here? Does one have TLS or something?

Comment (NickGerleman) [reply vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp:71]: I'm curious what's going on here more in general, with having the server running on a previous port and no longer calling Start/Stop on it.

Comment (JunielKatarn) [reply vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp:71]: No. Currently, the in-proc WebSocket server doesn't support multiple requests (yes, that's pretty limited). I deafult that server to port 5556 so it doesn't collide with the Node one provided by React Native.

For now, I'm using the Node server to run this test.

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp [hunk 2]
```diff
@@ -198,36 +209,37 @@ fire_and_forget WinRTWebSocketResource::PerformWrite()
...
       length = message.size();
       winrt::array_view<const uint8_t> arr(
         CheckedReinterpretCast<const uint8_t*>(message.c_str()),
         CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
-      m_writer.WriteBytes(arr);
+      self->m_writer.WriteBytes(arr);
     }
 
-    co_await m_writer.StoreAsync();
+    auto operation = m_writer.StoreAsync();
```
Comment (NickGerleman) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:230]: nit: `self->m_writer` for consistency. Though accessing raw member variables is safe here, so long as you have the strong ptr in scope. I.e. you can remove the `self->` entirely if you wanted to.

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp [hunk 3]
```diff
@@ -96,99 +98,108 @@ WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString, vector<C
...
 {
-  if (m_writeQueue.empty())
+  auto self = shared_from_this();
+  if (self->m_writeQueue.empty())
   {
+    self = nullptr;
     co_return;
   }
 
+  self = shared_from_this();
```
Comment (NickGerleman) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:185]: Why do we need this?

**Comment by NickGerleman** [comment]
Signing off, but I'm going to turn off AutoMerge first in case we want to address nits before checkin.

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp (hunk 3 above)
Comment (JunielKatarn) [reply vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:185]: We don't. Leftover from a `weak_ptr` prototype.
Removing.

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp (hunk 2 above)
Comment (JunielKatarn) [reply vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:229]: Noted.
I'll just add the missing `self->` to make the change simpler.
Will remove in a subsequent change.

Comment (JunielKatarn) [reply vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:229]: Another leftover from the `weak_ptr` attempt.
There is no need to store the action/operation in a variable anymore.

File: vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp (hunk 1 above)
Comment (JunielKatarn) [reply vnext/Desktop.IntegrationTests/WebSocketResourcePerformanceTests.cpp:71]: > I'm curious what's going on here more in general, with having the server running on a previous port and no longer calling Start/Stop on it.

It's only declared, but disabled (never runs).
I could as well just remove it and leave a comment to re-introduce it later.

**Comment by JunielKatarn** [comment]
Re-enabling `Automerge`, for the review feedback has been addressed.

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp [hunk 4]
```diff
@@ -345,11 +383,14 @@ void WinRTWebSocketResource::SendBinary(const string& base64String)
 
 void WinRTWebSocketResource::Close(CloseCode code, const string& reason)
 {
+  if (m_readyState == ReadyState::Closing || m_readyState == ReadyState::Closed)
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:386]: >if [](start = 2, length = 2)

Is this going to be accessed from single thread, or this api is multithreaded?

Comment (JunielKatarn) [reply vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:386]: It can potentially be accessed from multiple threads.
